lib/winapi/filemanip.c

```diff
@@ -241,6 +242,130 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
     }
 }
 
+BOOL CopyFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName, BOOL bFailIfExists)
+{
+    static const DWORD readBufferSize = 65535;
+    NTSTATUS status;
+    HANDLE sourceHandle;
+    HANDLE targetHandle = INVALID_HANDLE_VALUE;
+    ANSI_STRING sourcePath;
+    ANSI_STRING targetPath;
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_NETWORK_OPEN_INFORMATION networkOpenInformation;
+    LPVOID readBuffer;
+    DWORD bytesRead;
+
+    RtlInitAnsiString(&sourcePath, lpExistingFileName);
+    InitializeObjectAttributes(&objectAttributes, &sourcePath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(
+            &sourceHandle,
+            FILE_GENERIC_READ,
+            &objectAttributes,
+            &ioStatusBlock,
+            FILE_SHARE_READ,
+            FILE_READ_ATTRIBUTES | FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT | FILE_NO_EA_KNOWLEDGE);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    // Fail overwriting the file if the target == the source. This is
+    // intentionally done after the check for existence of the source
+    // to set the most appropriate LastError value.
+    if (!strcmp(lpExistingFileName, lpNewFileName)) {
+      NtClose(sourceHandle);
+      SetLastError(ERROR_SHARING_VIOLATION);
+      return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+            sourceHandle,
+            &ioStatusBlock,
+            &networkOpenInformation,
+            sizeof(networkOpenInformation),
+            FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtCreateFile(
+            &targetHandle,  // FileHandle
+            FILE_GENERIC_WRITE,   // DesiredAccess
+            &objectAttributes, // ObjectAttributes
+            &ioStatusBlock,   // IoStatusBlock
+            &networkOpenInformation.AllocationSize,  // AllocationSize
+            networkOpenInformation.FileAttributes,  // FileAttributes
+            FILE_SHARE_READ | FILE_SHARE_WRITE,  // ShareAccess
+            bFailIfExists ? FILE_CREATE : FILE_SUPERSEDE,  // CreateDisposition
+            FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT);  // CreateOptions
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    readBuffer = MmAllocateContiguousMemory(readBufferSize);
```

Comment:
Not sure if a contiguous memory allocation is a good idea here. We should reserve those for GPU resources which can't deal with paging?

Not sure what other parts of winapi do.

Reply:
I didn't see any other allocations in the winapi dir but happy to use whatever is most appropriate (I'm not very familiar w/ the NT kernel).

Reply:
Switched to `NtAllocateVirtualMemory`